gpu/octree/src/cuda/radius_search.cu

```diff
@@ -298,20 +315,45 @@ namespace pcl
                     const bool out_of_bounds = (offset + total_new) >= length_left;                              
 
                     if (take && !out_of_bounds)
+                    {
                         out[offset] = batch.indices[beg + idx];
+                        sqr_dist[offset] = d2;
+                    }
 
                     const int new_nodes = storage.cta_buffer[last_threadIdx];
 
                     idx += Warp::STRIDE;
 
                     total_new += new_nodes;
-                    out += new_nodes;                
+                    out += new_nodes;
+                    sqr_dist += new_nodes;
 
                     if (__all_sync(0xFFFFFFFF, idx >= length) || __any_sync(0xFFFFFFFF, out_of_bounds) || total_new == length_left)
                         break;
                 }
                 return min(total_new, length_left);
             }
+
+            __device__ void squareDistancesKernel(const int beg, const float3& active_query, const int length, float* sqr_dist)
+            {
+                unsigned int idx = Warp::laneId();
+                while (true)
+                {
+                    if (idx < length)
+                    {
+                        const float dx = batch.points.ptr(0)[beg + idx] - active_query.x;
+                        const float dy = batch.points.ptr(1)[beg + idx] - active_query.y;
+                        const float dz = batch.points.ptr(2)[beg + idx] - active_query.z;
+
+                        sqr_dist[idx] = dx * dx + dy * dy + dz * dz;
+                    }
+
+                    idx += Warp::STRIDE;
+
+                    if (__all_sync(0xFFFFFFFF, idx >= length))
+                        break;
+                }
+            }
```

Comment:
This should be converted into a for loop. Notice you have an initialization value, a condition you check every time and a variable increment.